pandas/io/parsers.py

```diff
@@ -1677,10 +1686,27 @@ def _infer_columns(self):
 
             columns = []
             for level, hr in enumerate(header):
-                line = self._buffered_line()
+                try:
+                    line = self._buffered_line()
 
-                while self.line_pos <= hr:
-                    line = self._next_line()
+                    while self.line_pos <= hr:
+                        line = self._next_line()
+
+                except StopIteration:
```

Comment:
How complicated is this code?  It's essentially boilerplate and logic from the C engine code.  I'm not writing anything new here.  I don't understand why people would have a hard time understanding this code but yet the C engine code is considered comprehensible.